### ISSUE TITLE
Add a better implementation of a file-backed multipart decoder

### ISSUE THREAD

**Issue by nigredo-tori** [http4s/http4s#2627]
`MultipartDecoder.mixedMultipart` is package-private, and for a good reason: it requires the user to drain all the decoded `Part`s - and each can only be read once! This is a rather ugly restriction, which stems from the fact that `EntityDecoder` abstraction doesn't provide any resource handling.

Rather than shoehorn resource handling into an abstraction that doesn't support it, we should make it explicit. Something like this would do:
```scala
def proposedMultipart(...): Resource[F, EntityDecoder[F, Multipart[F]]]
// Usage
proposedMultipart(...).use { decoder =>
  req.decodeWith(decoder) { multipart =>
    ???
  }
}
```
This gives the user a way to explicitly specifly the multipart lifetime, and to rely on the guarantees that `Resource` provides.

As for the actual resource handling, I imagine we can keep track of created files while parsing, and delete all of them when the `Resource` is released. Or we can just create a temporary directory on acquire, create files in it while parsing, and delete the directory on release. Both these solutions promise to simplify resource handling in the `MultipartParser` (the `cleanupFile` calls littering the file).

**Comment by rossabaker** [comment]
Would anything ever happen between the `.use` and the `req.decodeWith(decoder)`?  Do we gain much over `Resource[F, Multipart[F]]`?

**Comment by nigredo-tori** [comment]
> Would anything ever happen between the .use and the req.decodeWith(decoder)?

Probably not. We could add some sugar for that if we expect `Resource[F, EntityDecoder[F, A]]` to become a pattern. Though the same goes for the `EntityDecoder[F, Resource[F, A]]` approach.

> Do we gain much over Resource[F, Multipart[F]]?

1. The actual decoding happens where we expect it to (in `Request.decodeWith`), rather then when acquiring the resulting resource.
1. We avoid potential pitfalls of not running the resulting `Resource[F, Multipart[F]]`, running it too late, or running it twice somehow.
1. It works better with "bracketing" functions like `Request.decodeWith`, which might be useful once we begin `Resource`ifying the server (#2663).

**Comment by rossabaker** [comment]
Reading this again, I like it.

**Comment by fredshonorio** [comment]
It looks like this issue could be closed given that #4915 was merged.

**Comment by nigredo-tori** [comment]
> It looks like this issue could be closed given that #4915 was merged.

Absolutely. It probably was not automatically closed because `series/0.23` was not the default branch when the PR was being merged.
